Automattic/pocket-casts-ios
# Fix crash when selecting title of featured episode

Fixes #58 

This PR fixes a data issue that I haven't been able to reproduce in practice, but I have a hunch is occurring because it seems like the only hole is in the "Featured Episode" tap handler.

The issue stems from the fact that if the podcast has already been downloaded, the existing entry is being used instead of pulling new data. If the episode that is featured is a newer episode than what was downloaded the "featured episode" will be missing from the list and the view will crash trying to load the episode. Reasons why the episode might be missing could be the sync hasn't happened yet, or is happening but is slow due to slow network connection or the device being offline.

This PR adds a new optional parameter to the `loadPodcast` method to pass the episode that we are interested in, and if it doesn't exist, the podcast data will be refreshed. If it can't be refreshed or if the episode is still missing after it is refreshed, the episode view will not be opened.

These changes should prevent the crashes in the Discover tab, but has the weird user behaviour of tapping on an episode and having nothing happen. We could add some feedback here, however this is also currently the behaviour if you don't have internet access and you try to tap the episode when you don't already have the podcast itself downloaded.

There is a similar crash reported coming from the ListeningHistory view, but I suspect the cause of that one might be different even though the same child view is causing the crash.

## To test

1. Switch the Build configuration for the `pocketcasts` scheme to `Staging`, this will ensure you have a Featured Episode available to tap in the Discover view since the staging data has a few loaded
2. debug and in the simulator visit the Discover view, tap the Featured Episode title (**not** the "Play episode" button) and the episode should load as expected (ensuring everything still works correctly)
    
<img width="150px" src="https://user-images.githubusercontent.com/4081020/193080632-7cecf677-dfd8-4868-b48d-b94e2ceb35d7.png" />

3. Simulate the episode is missing from the app db:
   - add the following code at line ~~144~~  ~~148~~ 146 of `DiscoverEpisodeViewModel.swift` : `DataManager.sharedManager.delete(episodeUuid: episodeUuid) // testing line to exercise code block`
   - debug again, tap the Featured Episode title and the episode should load as expected (simulates downloading the episode when its missing)
   - turn off wifi, tap the Featured Episode title and nothing should happen (because the episode can't be downloaded) - but importantly the app shouldn't crash either
4. Simulate the episode is missing from app db AND is completely missing from the podcast (maybe it was deleted by the creator):
   - ~~replace line 191 with `let episode:Episode? = nil`~~
   - ~~replace line 192 with `completion?(false)`~~
   - replace line 169 with `promise(.failure(.episodeNotFound))`
   - debug again, tap the Featured Episode title and nothing should happen (because we're simulating the episode not existing anymore) - but importantly the app doesn't crash either

## Checklist


Tests were considered but the function I'd need to write tests for is `static func loadPodcast` which means I'd have to inject the mocked data manager into that function, unless there is a way to override the static `DataManager.sharedManager` during testing? I'm happy to write the tests if there is a clean way to test this function, or if we're fine with the messy way of adding the DataManager as a param to 2 static functions 😄
